File: 2025/day6/main.py
with open("input.txt", "r") as f:
    lines = [line.rstrip("\n") for line in f]

# Pad lines to same length
max_len = max(len(line) for line in lines) if lines else 0
lines = [line.ljust(max_len) for line in lines]

rows = len(lines)
cols = max_len

# Group columns into problems
problems = []
current_cols = []

for c in range(cols):
    # Check if column is all spaces
    is_separator = all(lines[r][c] == " " for r in range(rows))

    if is_separator:
        if current_cols:
            problems.append(current_cols)
            current_cols = []
    else:
        current_cols.append(c)

if current_cols:
    problems.append(current_cols)

# Solve each problem
grand_total = 0

for problem_cols in problems:
    numbers = []

    # Extract numbers (all rows except last)
    for r in range(rows - 1):
        num_str = "".join(lines[r][c] for c in problem_cols).strip()
        if num_str:
            numbers.append(int(num_str))

    # Extract operator (last row)
    operator = "".join(lines[rows - 1][c] for c in problem_cols).strip()

    # Calculate result
    if operator == "+":
        result = sum(numbers)
    else:  # operator == '*'
        result = 1
        for num in numbers:
            result *= num

    grand_total += result

print(grand_total)

File: 2025/day6/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	file, _ := os.Open("input.txt")
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	// Pad to same length
	maxLen := 0
	for _, line := range lines {
		if len(line) > maxLen {
			maxLen = len(line)
		}
	}
	for i := range lines {
		for len(lines[i]) < maxLen {
			lines[i] += " "
		}
	}

	rows := len(lines)
	cols := maxLen

	// Group columns into problems
	var problems [][]int
	var currentCols []int

	for c := 0; c < cols; c++ {
		isSeparator := true
		for r := 0; r < rows; r++ {
			if lines[r][c] != ' ' {
				isSeparator = false
				break
			}
		}

		if isSeparator {
			if len(currentCols) > 0 {
				problems = append(problems, currentCols)
				currentCols = nil
			}
		} else {
			currentCols = append(currentCols, c)
		}
	}
	if len(currentCols) > 0 {
		problems = append(problems, currentCols)
	}

	// Solve each problem
	grandTotal := 0

	for _, problemCols := range problems {
		var numbers []int

		// Extract numbers
		for r := 0; r < rows-1; r++ {
			var sb strings.Builder
			for _, c := range problemCols {
				sb.WriteByte(lines[r][c])
			}
			numStr := strings.TrimSpace(sb.String())
			if numStr != "" {
				num, _ := strconv.Atoi(numStr)
				numbers = append(numbers, num)
			}
		}

		// Extract operator
		var opSb strings.Builder
		for _, c := range problemCols {
			opSb.WriteByte(lines[rows-1][c])
		}
		operator := strings.TrimSpace(opSb.String())

		// Calculate
		result := 0
		if operator == "+" {
			for _, num := range numbers {
				result += num
			}
		} else {
			result = 1
			for _, num := range numbers {
				result *= num
			}
		}

		grandTotal += result
	}

	fmt.Println(grandTotal)
}
